src/terminal/adapter/adaptDispatchGraphics.cpp

```diff
@@ -333,213 +43,227 @@ static constexpr bool _isDefaultColorOption(const DispatchTypes::GraphicsOptions
-                                        COLORREF& rgbColor,
-                                        bool& isForeground,
-                                        size_t& optionsConsumed)
+// - The number of options consumed, not including the initial 38/48.
+size_t AdaptDispatch::_SetRgbColorsHelper(const std::basic_string_view<DispatchTypes::GraphicsOptions> options,
+                                          TextAttribute& attr,
+                                          const bool isForeground)
 {
-    bool success = false;
-    optionsConsumed = 1;
-    if (options.size() >= 2 && _isRgbColorOption(til::at(options, 0)))
+    size_t optionsConsumed = 0;
+    if (options.size() >= 1)
     {
-        optionsConsumed = 2;
-        const auto extendedOpt = til::at(options, 0);
-        const auto typeOpt = til::at(options, 1);
-
-        if (extendedOpt == DispatchTypes::GraphicsOptions::ForegroundExtended)
-        {
-            isForeground = true;
-        }
-        else if (extendedOpt == DispatchTypes::GraphicsOptions::BackgroundExtended)
-        {
-            isForeground = false;
-        }
-
-        if (typeOpt == DispatchTypes::GraphicsOptions::RGBColorOrFaint && options.size() >= 5)
+        optionsConsumed = 1;
+        const auto typeOpt = til::at(options, 0);
+        if (typeOpt == DispatchTypes::GraphicsOptions::RGBColorOrFaint && options.size() >= 4)
         {
-            optionsConsumed = 5;
+            optionsConsumed = 4;
+            const size_t red = til::at(options, 1);
+            const size_t green = til::at(options, 2);
+            const size_t blue = til::at(options, 3);
             // ensure that each value fits in a byte
-            unsigned int red = std::min(static_cast<unsigned int>(til::at(options, 2)), 255u);
-            unsigned int green = std::min(static_cast<unsigned int>(til::at(options, 3)), 255u);
-            unsigned int blue = std::min(static_cast<unsigned int>(til::at(options, 4)), 255u);
-
-            rgbColor = RGB(red, green, blue);
-
-            success = _pConApi->SetConsoleRGBTextAttribute(rgbColor, isForeground);
+            if (red <= 255 && green <= 255 && blue <= 255)
+            {
+                const COLORREF rgbColor = RGB(red, green, blue);
+                attr.SetColor(rgbColor, isForeground);
+            }
         }
-        else if (typeOpt == DispatchTypes::GraphicsOptions::BlinkOrXterm256Index && options.size() >= 3)
+        else if (typeOpt == DispatchTypes::GraphicsOptions::BlinkOrXterm256Index && options.size() >= 2)
         {
-            optionsConsumed = 3;
-            if (til::at(options, 2) <= 255) // ensure that the provided index is on the table
+            optionsConsumed = 2;
+            const size_t tableIndex = til::at(options, 1);
+            COLORREF rgbColor;
+            if (_pConApi->PrivateGetColorTableEntry(tableIndex, rgbColor))
             {

```

Comment:
```suggestion
            {
                // TODO GH#xxxx: Decouple xterm-256color indexed storage from RGB storage
```

(I don't know if this is best covered by 2661.)

Reply:
Yeah, this can't be done until the color tables are unified (essentially issue #1223). I've got a followup PR in the works that will address this, but I'll add a TODO comment in the meantime.